### DISCUSSION TITLE
Using Symfony dedicated Slugger

### DISCUSSION THREAD

**Discussion by ghost** [thephpleague/commonmark#860] [Ideas]
Hi,

I was glad this week-end when I found your bundle to convert Markdown to HTML with all your extensions.

But I was surprised when I tried the HeadingPermalinkExtension combined with the TableOfContentsExtension; I'm french and the generated slugs keep accents and do not replace apostrophe as expected (empty string instead of '-').

I was wondering if you could implement directly the new dedicated component [Symfony\Component\String\Slugger\AsciiSlugger](https://symfony.com/doc/current/components/string.html#slugger)?

This would ensure coherence over applications.

Moreover, that would be great if we could pass a callback for formating it. For example, in my case, my titles got some content between parentheses but I don't want the slugs to reflect this information and I'd like to strip it when generating slug. That would be a cool option too. :-)

Thank for your answer about it.

**Comment by colinodell** [comment]
Hi @fanchyfanch! Thanks for your kind feedback :) [That is currently customizable](https://commonmark.thephpleague.com/2.3/customization/slug-normalizer/) - I think we're just missing a link to this information from the HeadingPermalink extension docs.

**Reply by ghost** [reply]
No no, that's OK, I've read this documentation last week-end and was able to configure it as it is indicated. 

But generated slug for "_Création d’un groupe par défaut_" was **#création-dun-groupe-par-défaut**.

See? I still got **accents** and **apostrophe** wasn't strip right.

Actually I'm talking about the **League\CommonMark\Normalizer\SlugNormalizer**:

```PHP
public function normalize(string $text, array $context = []): string
    {
        // Add any requested prefix
        $slug = ($context['prefix'] ?? '') . $text;
        // Trim whitespace
        $slug = \trim($slug);
        // Convert to lowercase
        $slug = \mb_strtolower($slug);
        // Try replacing whitespace with a dash
        $slug = \preg_replace('/\s+/u', '-', $slug) ?? $slug;
        // Try removing characters other than letters, numbers, and marks.
        $slug = \preg_replace('/[^\p{L}\p{Nd}\p{Nl}\p{M}-]+/u', '', $slug) ?? $slug;
        // Trim to requested length if given
        if ($length = $context['length'] ?? $this->defaultMaxLength) {
            $slug = \mb_substr($slug, 0, $length);
        }

        return $slug;
    }
```

while you coud do: 

```PHP
public function normalize(string $text, array $context = []): string
    {
        // Add any requested prefix
        $slug = ($context['prefix'] ?? '') . $text;
        // Trim whitespace
        $slug = \trim($slug);
        // Convert to lowercase
        $slug = \mb_strtolower($slug);
        // Use Symfony internal slugger 
        $slug = (new Symfony\Component\String\Slugger\AsciiSlugger())->slug($slug)->__toString();
        // Trim to requested length if given
        if ($length = $context['length'] ?? $this->defaultMaxLength) {
            $slug = \mb_substr($slug, 0, $length);
        }

        return $slug;
    }
```

And the result is properly processed to: **#creation-d-un-groupe-par-defaut**. :-)
